### PR TITLE
Python3

### PR DESCRIPTION
**Pull request by Thisisnotdalton**
Separating some of the main GillesPy code to individual modules. Some Python3 fixes.

### CODE REVIEW

**Comment by CrookedWay** [comment]
This looks really good Dalton! I'm really glad you fixed all of the import statements. 

Are there benchmarks for the difference between the implementation of your optimized SSA and the previous iteration? 

I'm a little bit curious (And this is by definition more of a version support question/dogmatic question) if it would be be better to use [PEP-0484](https://www.python.org/dev/peps/pep-0484/) style for the type error handling. It would be more compact and would perhaps be more dogmatic for future students on the project who are more used to more explicit typing.

**Comment by Thisisnotdalton** [comment]
There are benchmarks I had exported to .png images listed as SSA_Comparisons. The original Python implementation we had is listed as the Python entry. The model was still somewhat simplified for the tests, but with few changes to the algorithms other than preallocating memory and making greater use of Numpy and Cython, I would imagine it is still sufficient.

As for the PEP-0484 style, I think that that sounds like a better approach to the type errors. I am still new to standardizations such as that, but I trust your input.

File: gillespy2/basic_ssa_solver.py [hunk 1]
```diff
@@ -12,64 +12,59 @@ def run(self, model, t=20, number_of_trajectories=1,
...
+                tau = -1*math.log(random.random())/prop_sum
+                curr_time += tau
+                while(curr_time > save_time and curr_time <= t):
+                    results['time'].append(save_time)
+                    for s in model.listOfSpecies:
+                        results[s].append(curr_state[s])
+                    save_time += increment	
 
+                for react in model.listOfReactions[reaction].reactants:
+                    curr_state[str(react)] -=  model.listOfReactions[reaction].reactants[react]
```
Comment (NBreedlove) [inline gillespy2/basic_ssa_solver.py]: Real changes here line 63,65

File: gillespy2/gillespy2.py [hunk 2]
```diff
@@ -391,7 +391,7 @@ def run(self, number_of_trajectories=1, seed=None,
             Use names of species as index of result object rather than position numbers.
```
Comment (NBreedlove) [inline gillespy2/gillespy2.py]: and here. Looks good, streamlined the string index

**Comment by CrookedWay** [comment]
So, I `git checkout Python3` to start porting some of the graphing function over and I noticed that I can't get the library to run anymore on the standard Michaelis–Menten ipython notebook. I get an error pointing to the importation of the GilespyError submodule. 

I spun up a venv and an AWS EC2 instance to check to make sure that the problem isn't machine dependent, and it's the problem persists. I've been looking it over for like an hour and I can't see why it has issues finding or porting the submodule in. (It can find the other files in the directory so I'm fairly certain it's not a sys append issue.) 

Removing the import call fixes the issue, but when it tries to run the Exception Classes it fails to compile/run/do the math thing.

Nathan said earlier today that he had the same (or similar) problem and was going to speak to Dr. Drawert about it. I don't know what's going on with that.

I think what I'm going to do is create a branch for the style change I was talking about and merge the gillespy error module into main. I understand that it's commonly held best practice<sup>[1](#1)</sup> to have a seperate exception submodule, but when it's six or so lines it seems kind of extraneous to have an extra file.

<a name="1">1</a>: [Official Python Docs](https://docs.python.org/3/tutorial/errors.html), [Some dude's blog.](https://julien.danjou.info/blog/2016/python-exceptions-guide)

**Comment by CrookedWay** [comment]
I have a more robust modification (Formatting and Error checking), but if we want to just go with [this](https://github.com/GillesPy2/GillesPy2/tree/FixedImport)

**Pull request merged**